### DISCUSSION TITLE
PE and SIMD values

### DISCUSSION THREAD

**Discussion by sakshigoel155** [Xilinx/finn#489] [Q&A] [answered]
Hi,

After doing the FINN tutorials, I am now trying to create my own neural network for image classification. I have two questions about network parameters for deploying the model to PYNQ. 

1. I have trained the pytorch model with input [64,1,28,28] ([BS,C,H,W]). Now when i generate onnx transformation for this model, using FINN manager, should the input here have batchsize used in training i.e 64 or always1? 
FINNManager.export(fc, input_shape=(**1 or 64?**, 1, 28, 28), export_path='finn_FMNIST_1.onnx')
I saw in end2end examples, it was 1, so i want to know if batch_size is counted towards PE or SIMD calculations if it is not 1.

2. The second question is about PE and SIMD values itself. Now I read, that for each layer output%PE==0 and Input%SIMD==0. I used divisible values but ZynqBuild given an error. 
Can someone please explain this error and point me to some examples of customizing parallelization parameters. (including in/outFifoDepth)  
Attached: Error, PE,SIMD values used and convolution layer parameter

Any help in improving my understanding while I try to figure it out will be very helpful
Regards,
Sakshi 
![1](https://user-images.githubusercontent.com/74159729/148960078-e355a2c5-9b36-477c-abaf-4785e72649eb.png)
![2](https://user-images.githubusercontent.com/74159729/148960083-83a644bd-e084-40b3-8ad4-8e528afa66aa.png)
![3](https://user-images.githubusercontent.com/74159729/148960088-0789a814-8602-4cab-b066-509f9d8f8358.png)

**Comment by shashwat1198** [comment]
Just to add to the above query.

What are the default values of PE, SIMD & fifo depth if the user does not define any?

**Reply by sakshigoel155** [reply]
Hi,
I am not sure if this applies, but I saw this in the read docs. Maybe you can run buildconfig and confirm.
![pe_simd_def](https://user-images.githubusercontent.com/74159729/149747587-f50030db-ea87-4f2c-a02a-73cb87d8b0ef.png)
Also, if you see the onnx model before updating PE and SIMD, the values are set to 1 for all StreamingFCLayer_Batch. So default seems to be 1 to me, but I have just started using FINN and am not sure if I am missing something.
Further, according to cybersecurity tutorial, if you do define a target_fps, then compiler decides resource utilization on its own to match that target.

**Reply by shashwat1198** [reply]
Hello,
I am also in the stage of exploring finn. I found when you run the build dataflow for resource estimates then there is a file called auto_folding_config.json where you can see the PE and SIMD values that the compiler assigns to your model.

Thank you for your response,
Shashwat

**Comment by rpitonak** [comment]
Hello,

Some intuition to your second question
> The second question is about PE and SIMD values itself. Now I read, that for each layer output%PE==0 and Input%SIMD==0. I used divisible values but ZynqBuild given an error.
Can someone please explain this error and point me to some examples of customizing parallelization parameters. (including in/outFifoDepth)
Attached: Error, PE,SIMD values used and convolution layer parameter

It is important to notice that output and input are in this case meant to be  `channels` or `filters`. For example, if you have an input to the first layer `224x224x3`(RGB image) you can set the SIMD value to `3` or `1` so channels are divisible by it.

PE then depends on the output filters from the layer, e.g for 256 output from the layer in `QuantLinear` you can set PE to `256,128... etc`.

Hope it helps, let me know if something is unclear.

**Reply by sakshigoel155** [reply]
Hi,

Thank you for your reply. Appreciate it.
But I have a further question about SIMD. You mentioned it should be channels or filters but if I see in the ONNX graph for tfc_end2end_example, for the first StreamingFCLayer_Batch, the input is (1X784) from (1,1,28,28) input where channel is 1 and no kernel is used i think. But SIMD is set to 49 which is multiple of 784. So I am still not sure how to get a correct SIMD value. I understand PE assignment thanks.
I tried a build with SIMD=1 (as my input was 1 channel) and PE%output and it kept on running forever. I am wondering if there is another way because ZynQBuild for end2end examples was done in half hour approx.
![simd](https://user-images.githubusercontent.com/74159729/149758858-af8d7ccb-a549-4d37-9ccd-2073121f53c9.png)

Regards,
Sakshi

**Reply by dj-park** [reply, marked as answer]
As this is already 1 year old, you must've already figured it out. But for those who have similar questions... think of **MW%SIMD==0 and MH%PE==0** when MH is the weight matrix's height and MW is the weight matrix's width. 

In the case of fully connected layer (screenshot's case), SIMD values can be divisors of 784(MW) and PE values can be divisors of 64(MH). 

In CNN, you know convolution can be translated into matrix multiplication. In this case also, **MW%SIMD==0 and MH%PE==0**. Because convolution is equivalent to _N X (C(K^2))_ matrix multiplied by _(C(K^2)) X HW_ matrix (N: num of  output ch, C: num of input ch, K: kernel size, H: input H, W: input W), MW is (C(K^2) and MH is N. Thus, **C%SIMD==0 and N%PE==0**.
